File: sample/src/main/java/com/facebook/samples/litho/NavigatableDemoActivity.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.samples.litho;

import android.content.Intent;
import android.os.Bundle;
import android.view.MenuItem;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.NavUtils;
import java.util.Arrays;

public class NavigatableDemoActivity extends AppCompatActivity {

  private Demos.DemoItem mDataModel;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    int[] indices = getIntent().getIntArrayExtra(DemoListActivity.INDICES);
    if (indices != null) {
      mDataModel = getDataModel(indices);
      getSupportActionBar().setDisplayHomeAsUpEnabled(true);
      setTitle(mDataModel.getName());
    }
  }

  @Override
  protected void onDestroy() {
    super.onDestroy();
    mDataModel = null;
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
        // Respond to the action bar's Up/Home button
      case android.R.id.home:
        NavUtils.navigateUpFromSameTask(this);
        return true;
    }
    return super.onOptionsItemSelected(item);
  }

  @Override
  public @Nullable Intent getParentActivityIntent() {
    int[] indices = getIntent().getIntArrayExtra(DemoListActivity.INDICES);
    if (indices == null) {
      return null;
    }

    // Keep popping indices off the route until we get to an Activity we can navigate back to.
    // If we can't find one, pop back to the root.
    int[] parentIndices = indices;
    Demos.DemoItem item = null;
    while (parentIndices.length > 1 && !(item instanceof Demos.NavigableDemoItem)) {
      parentIndices = Arrays.copyOf(indices, parentIndices.length - 1);
      item = getDataModel(parentIndices);
    }

    if (item != null && item instanceof Demos.NavigableDemoItem) {
      return ((Demos.NavigableDemoItem) item).getIntent(this, parentIndices);
    } else {
      return new Intent(this, SampleRootActivity.class);
    }
  }

  private Demos.DemoItem getDataModel(int[] indices) {
    Demos.DemoItem model = Demos.DEMOS.get(indices[0]);
    for (int i = 1; i < indices.length; i++) {
      if (model instanceof Demos.HasChildrenDemos) {
        model = ((Demos.HasChildrenDemos) model).getDemos().get(indices[i]);
      } else {
        throw new RuntimeException("Unexpected type: " + model);
      }
    }
    return model;
  }

  protected Demos.DemoItem getDataModel() {
    return mDataModel;
  }
}

File: sample-kotlin/src/main/kotlin/com/facebook/samples/litho/kotlin/NavigatableDemoActivity.kt
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.samples.litho.kotlin

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NavUtils
import com.facebook.samples.litho.kotlin.demo.DataModels
import com.facebook.samples.litho.kotlin.demo.DemoListActivity
import com.facebook.samples.litho.kotlin.demo.DemoListActivity.Companion.INDICES
import com.facebook.samples.litho.kotlin.demo.DemoListDataModel
import java.util.Arrays

@SuppressLint("Registered")
open class NavigatableDemoActivity : AppCompatActivity() {

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    intent.getIntArrayExtra(INDICES).takeIf { it != null }?.let {
      supportActionBar?.setDisplayHomeAsUpEnabled(true)
      setTitleFromIndex(it)
    }
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean =
      if (item.itemId == android.R.id.home) {
        NavUtils.navigateUpFromSameTask(this)
        true
      } else {
        super.onOptionsItemSelected(item)
      }

  override fun getParentActivityIntent(): Intent? {
    val indices = intent.getIntArrayExtra(DemoListActivity.INDICES) ?: return null

    return Intent(this, DemoListActivity::class.java).also {
      if (indices.size > 1) {
        it.putExtra(DemoListActivity.INDICES, Arrays.copyOf(indices, indices.size - 1))
      }
    }
  }

  private fun setTitleFromIndex(indices: IntArray) {
    var dataModels: List<DemoListDataModel> = DataModels.DATA_MODELS

    for (i in 0 until indices.size - 1) {
      dataModels = dataModels[indices[i]].datamodels ?: dataModels
    }

    dataModels[indices[indices.size - 1]].also { title = it.name }
  }
}
